### PR TITLE
Reduce request latency from the agent to the proxy

### PR DESCRIPTION
**Pull request by ojarjur**
This change makes two improvements to the latency of requests sent
from the agent to the proxy:

1. Make the client timeout configurable. This fixes a performance
   issue where calls to a proxy with a longer timeout cause the
   agent's request to timeout, which in turn causes the agent to back
   off and delay subsequet requests to the proxy. The timeout needs to
   be configurable so that it can be made to match the proxy.
2. Don't make requests to the proxy block on reading a VM identity
   token from the metadata server. The requests to read VM identity
   seem to take up to 100ms, so by fetching the VM identity out of
   band, we can eliminate a long poll from requests to the proxy.

My exceedingly unscientific testing suggests that this drops the average
round-trip latency per request from ~150ms to ~110ms.

### CODE REVIEW

File: agent/utils/utils.go [hunk 1]
```diff
@@ -140,29 +136,73 @@ func hasVMServiceAccount() bool {
...
+	wrapped  http.RoundTripper
+	idChan   chan string
 }
 
-// addVMIDHeader adds a header to the given request identifying the VM (if any) on which the agent is running.
+func (t *vmTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	select {
+	case <-r.Context().Done():
+		return nil, fmt.Errorf("request cancelled")
+	case id := <-t.idChan:
```
Comment (abdallag) [inline agent/utils/utils.go]: This means, if the fe client request rate is higher than the ticker rate ( which is currently set to 10 seconds), the client would experience up to 10 seconds delays.

I would use a simple variable for caching here and not a channel. You can wait on the initial value and use a way higher refresh rate. The token expires in order of multiple minutes as far as I recall.

**Comment by abdallag** [comment]
The idea not to block on getting the VM ID is a good one and could definitely improve performance. But I would not always wait on the channel and use a single variable cached version instead.

File: agent/utils/utils.go (hunk 1 above)
Comment (ojarjur) [reply agent/utils/utils.go]: > This means, if the fe client request rate is higher than the ticker rate ( which is currently set to 10 seconds), the client would experience up to 10 seconds delays.

No it doesn't.

The idChan here is constantly filled up regardless of whether or not the ticker ticks. The 10 second ticker rate just affects how frequently the value is updated, not how frequently it is sent on the channel.

That being said, I will try out the approach of caching a value (along with the required locking) and see if I like that better.

**Comment by ojarjur** [comment]
> The idea not to block on getting the VM ID is a good one and could definitely improve performance. But I would not always wait on the channel and use a single variable cached version instead.

I tried out the single-cached-value-with-locks alternative and decided I do like it better.

PTAL

**Pull request merged**